reactor-core/src/main/java/reactor/core/publisher/FluxLimitRate.java

```diff
@@ -0,0 +1,404 @@
+            final ConditionalSubscriber<? super T> cs =
+                (ConditionalSubscriber<? super T>) actual;
+          return new LimitRateConditionalSubscriber<>(cs, highTide, lowTide);
+        }
+        return new LimitRateSubscriber<>(actual, highTide, lowTide);
+    }
+
+    private final static class LimitRateSubscriber<T> implements InnerOperator<T, T> {
+
+        private final int highTide;
+        private final int limit;
+
+        volatile long requested; // need sync
+        static final AtomicLongFieldUpdater<LimitRateSubscriber> REQUESTED =
+            AtomicLongFieldUpdater.newUpdater(LimitRateSubscriber.class, "requested");
+
+        volatile int wip; // need sync
+        static final AtomicIntegerFieldUpdater<LimitRateSubscriber> WIP =
+            AtomicIntegerFieldUpdater.newUpdater(LimitRateSubscriber.class, "wip");
+
+        volatile int produced = 0;
+        static final AtomicIntegerFieldUpdater<LimitRateSubscriber> PRODUCED =
+            AtomicIntegerFieldUpdater.newUpdater(LimitRateSubscriber.class, "produced");
+
+        private int pending;
+        // need sync since should be checked/zerroed in onNext
+        // and increased in request
+        private int counter; // no need to sync since increased zerroed only in
+        // the request method
+
+        final CoreSubscriber<? super T> actual;
+
+        Subscription s;
+
+        private LimitRateSubscriber(CoreSubscriber<? super T> actual,
+            int highTide,
+            int lowTide) {
+            this.actual = actual;
+            this.highTide = highTide;
+            this.limit = Operators.unboundedOrLimit(highTide, lowTide);
+        }
+
+        @Override
+        public void request(long n) {
+            if (Operators.validate(n)) {
+                Operators.addCap(REQUESTED, this, n);
+
+                tryRequest();
+            }
+        }
+
+        void tryRequest() {
+            if (WIP.getAndIncrement(this) != 0) {
+                return;
+            }
+
+
+            final Subscription s = this.s;
+            final int highTide = this.highTide;
+
+            int missed = 1;
+            int pending = this.pending;
+            int produced = this.produced;
+            long r = this.requested;
+            long deliver = 0;
+
+            for (;;) {
+                if (r != Long.MAX_VALUE || highTide != Integer.MAX_VALUE) {
+                    if (produced != 0) {
+                        deliver = Math.min(produced, r);
+                    }
+                    else if (pending != highTide) {
+                        deliver = Math.min(highTide - pending, r);
+                    }
+
+                    int w = wip;
+                    if (missed == w) {
+                        if (deliver > 0) {
+                            if (produced != 0) {
+                                this.pending = highTide - produced + (int) deliver;
```

Comment:
The many variables makes the logic hard to follow here.

Let's say we've got `produced == 12, r == 100`. That would mean `deliver == min(12, 100) == 12 == produced`. So here `pending = highTide - 12 + 12`.

Similarly below, if we reach line 132 it means `deliver > 0 && produced == 0`, so `deliver == highTide - pending` and thus line 132 boils down to `this.pending = pending + hightTide - pending = highTide`. So these computations only really make sense when `deliver == r` (ie r is smaller than the produced/hightide-pending).

Maybe that case of `deliver == r` should stand out a bit more? Something like:

```java
if (deliver > 0) {
  if (deliver == r) {
    int smallRequest = deliver; //for readability only, could be replaced with `deliver` + a comment
    if (produced != 0) {
      this.pending = highTide - produced + smallRequest;
      PRODUCED.addAndGet(this, -produced);
    }
    else {
      this.pending += smallRequest;
    }
  }
  else {
    this.pending = highTide;	
    if (produced != 0) {
      PRODUCED.addAndGet(this, -produced);
    }
  }
  Operators.addCap(REQUESTED, this, -deliver);

  missed = WIP.addAndGet(this, -missed);
  if (missed == 0) {
    s.request(deliver);
    break;
  }
}
```